modules/library/metadata/src/main/java/org/geotools/factory/CategoryRegistry.java

```diff
@@ -0,0 +1,319 @@
+		return streamCategories()
+				.filter(category -> category.isAssignableFrom(type))
+				.map(this::instanceRegistry)
+				.flatMap(registry -> stream(registry.getInstanceOfType(type)))
+				.findFirst();
+	}
+
+	/**
+	 * Orders the specified instances, so that the first appears before the second when
+	 * {@link #streamInstances(Class, boolean) iterateInstances} is called with
+	 * {@code useOrder = true}.
+	 *
+	 * @param category The category to order instances for.
+	 * @return {@code true} if this establishes a new order
+	 */
+	public <T> boolean setOrder(Class<T> category, T firstInstance, T secondInstance) {
+		ensureArgumentNonNull("firstInstance", firstInstance);
+		ensureArgumentNonNull("secondInstance", secondInstance);
+		return instanceRegistry(category).setOrder(firstInstance, secondInstance);
+	}
+
+	/**
+	 * Removes the ordering between the specified instances, so that the first no longer appears
+	 * before the second when {@link #streamInstances(Class, boolean) iterateInstances} is
+	 * called with {@code useOrder = true}.
+	 *
+	 * @param category The category to clear instance order for.
+	 * @return {@code true} if that ordering existed before
+	 */
+	public <T> boolean clearOrder(Class<T> category, T firstInstance, T secondInstance) {
+		ensureArgumentNonNull("firstInstance", firstInstance);
+		ensureArgumentNonNull("secondInstance", secondInstance);
+		return instanceRegistry(category).clearOrder(firstInstance, secondInstance);
+	}
+
+	private static class InstanceRegistry<T> {
+
+		private final FactoryRegistry factoryRegistry;
+		private final Class<?> category;
+
+		private final Map<Class<?>, T> instancesByType = new HashMap<>();
+		private final PartiallyOrderedSet<T> instances = new PartiallyOrderedSet<>(false);
+
+		private InstanceRegistry(FactoryRegistry factoryRegistry, Class<?> category) {
+			this.factoryRegistry = factoryRegistry;
+			this.category = category;
+		}
+
+		public boolean register(final T instance) {
+			ensureArgumentNonNull("instance", instance);
+			boolean deregistered = deregisterByType(instance);
+			registerInternal(instance);
+			notifyRegistered(instance);
+			return !deregistered;
+		}
+
+		private void notifyRegistered(final T instance) {
+			if (instance instanceof RegistrableFactory) {
+				((RegistrableFactory) instance).onRegistration(factoryRegistry, category);
+			}
+		}
+
+		private void registerInternal(final T instance) {
+			instancesByType.put(instance.getClass(), instance);
+			instances.add(instance);
+		}
+
+		public boolean deregister(final T instance) {
+			ensureArgumentNonNull("instance", instance);
+			if (instancesByType.containsKey(instance.getClass())) {
+				deregisterByType(instance);
+				return true;
+			} else {
+				return false;
+			}
+		}
+
+		// documentation hint: returns true if an instance of the same type was previously registered
+		private boolean deregisterByType(final T instance) {
+			T removed = instancesByType.remove(instance.getClass());
```

Comment:
The current code is perfectly safe, because `PartiallyOrderedSet::remove` and `notifyDeregistered` handle `null` well. But the `null` check could guard against future implementation errors, so it does make some sense. Do you want me to add it?